### PR TITLE
Make RandomWindowGeoDataset more efficient when sampling chips from scenes with AOIs

### LINKED ISSUE #1219: RandomWindowGeoDataset should be more efficient when sampling chips from a scene with AOIs
Currently, `RandomWindowGeoDataset` does a kind of rejection sampling, where it samples chips randomly from anywhere in the scene and then checks to see if they fall within the AOIs, and if not, it retries a fixed number of times and then gives up:
https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/dataset/dataset.py#L338-L357

This is very inefficient when the AOIs are scattered throughout the scene and make up only a small proportion of the total area.

Instead, it could just sample from within the AOI. In general, i.e for arbitrary AOIs, this is a non-trivial problem. Some starting points:
- https://mathematica.stackexchange.com/questions/57938/how-to-generate-random-points-in-a-region
- https://gis.stackexchange.com/questions/6412/generate-points-that-lie-inside-polygon
- http://kuanbutts.com/2017/08/17/delaunay-triangulation/

### PR DESCRIPTION
**Pull request by AdeelH**
## Overview

Makes the sampling of random windows from scenes with AOIs more efficient. The previous approach was to sample a random window from anywhere within the scene and check to see if it falls within the AOI. In contrast, this approach is able to generate random window locations that are always inside the AOI. We still need to check if the window is fully contained within the AOI, since sampled windows near the boundaries of the AOI will occasionally overflow the AOI.

The algorithm can be summarized as:
- Partition the AOI polygons into triangles (triangulation).
    - This procedure is also able to account for holes inside polygons (see example below).
- When sampling, pick a triangle (`ABC`) at random with probability proportional to its area.
- Starting at `A`, travel a random distance along vectors `AB` and `AC`.
- Return the final position.

#### Example AOI
aoi

#### Old sampling approach
random_chips

##### triangulation
triangulation
##### sampled windows
aoi_chips

### Checklist

- [x] Added `needs-backport` label if PR is bug fix that applies to previous minor release
- [x] Ran scripts/format_code and committed any changes
- [x] Documentation updated if needed
- [x] PR has a name that won't get you publicly shamed for vagueness

### Notes

- New dependency added: [triangle](https://github.com/drufat/triangle)

## Testing Instructions
- Rebuild docker image to install new dependency.
- Run an example that has an AOI, while using a `GeoDataWindowConfig` with `method=GeoDataWindowMethod.random`.

Closes #1219

### CODE REVIEW

**Comment by jamesmcclain** [comment]
:tada:

File: rastervision_pytorch_learner/rastervision/pytorch_learner/learner_config.py
Comment (AdeelH) [inline rastervision_pytorch_learner/rastervision/pytorch_learner/learner_config.py:470]: This is mostly here to provide a way to turn this functionality off on the off chance we encounter an input that the triangulation algorithm cannot handle.

Comment (AdeelH) [reply rastervision_pytorch_learner/rastervision/pytorch_learner/learner_config.py:470]: Oops. The 's' at the end is a typo. This supposed to be called `efficient_aoi_sampling` not `efficient_aoi_samplings`.

File: tests/pytorch_learner/dataset/test_aoi_sampler.py
Comment (lewfish) [inline tests/pytorch_learner/dataset/test_aoi_sampler.py:62]: Nice use of a hypothesis test.

**Review by lewfish — APPROVED** [review]
Nice addition! This is approved assuming you manually ran the style tests, unit tests, and integration tests locally, since CI is currently not running (but see https://github.com/azavea/raster-vision/issues/1220) Note that the semseg integration tests are failing https://github.com/azavea/raster-vision/issues/1220

**Comment by AdeelH** [comment]
@lewfish One important change since the review:

I've reverted the commit that applied the inverse affine transform in the GeoDataset. The reason that was needed in the first place was that `RasterioCRSTransformer.from_dataset` used to return an `IdentityCRSTransformer` when the `.tif` did not have a CRS even if it did have an affine transform (which was the case with the other project's data). Once I modified `RasterioCRSTransformer.from_dataset` to handle the case of no-CRS + non-identity-transform, everything else just started working as expected.

I'm okay with waiting till the CI is fixed before merging this.

**Comment by lewfish** [comment]
Can you do something to trigger CI on this PR? It looks like you force pushed yesterday but I'm not sure if that was after the CI PR was merged.

**Pull request merged**
